storage_controller/src/tenant_shard.rs

```diff
@@ -979,24 +980,52 @@ impl TenantShard {
                         ),
                     )
                 })
-                .collect::<Vec<_>>();
+                .collect::<HashMap<_, _>>();
 
             if secondary_scores.iter().any(|score| score.1.is_none()) {
-                // Don't have full list of scores, so can't make a good decision about which to drop unless
-                // there is an obvious one in the wrong AZ
-                for secondary in self.intent.get_secondary() {
-                    if scheduler.get_node_az(secondary) == self.intent.preferred_az_id {
+                // Trivial case: if we only have one secondary, drop that one
+                if self.intent.get_secondary().len() == 1 {
+                    return Some(ScheduleOptimization {
+                        sequence: self.sequence,
+                        action: ScheduleOptimizationAction::RemoveSecondary(
+                            *self.intent.get_secondary().first().unwrap(),
+                        ),
+                    });
+                }
+
+                // Try to find a "good" secondary to keep, without relying on scores (one or more nodes is in a state
+                // where its score can't be calculated), and drop the others.  This enables us to make progress in
+                // most cases, even if some nodes are offline or have scheduling=pause set.
+
+                // If we're in secondary mode, then we want to retain the secondary in our preferred AZ.  Otherwise
+                // we want to retain the secondary in a non-preferred AZ (a good failover location).
+                let want_preferred_az = matches!(self.policy, PlacementPolicy::Secondary);
+
+                if let Some(retain_secondary) = self.intent.get_secondary().iter().find(|n| {
+                    (scheduler.get_node_az(n) != self.intent.preferred_az_id) ^ want_preferred_az
+                        && secondary_scores
+                            .get(n)
+                            .expect("Built from same list of nodes")
+                            .is_some()
```

Comment:
nit: I find this easier to read.

```suggestion
                    let online =  secondary_nodes.get(n).expect("Built from same list of nodes").is_some();
                    let in_preferred_az = scheduler.get_node_az(n) == self.intent.preferred_az_id;
                    
                    online && want_preferred_az == in_preferred_az
```

Reply:
Similar change in 711abf703ee0ce5425e8d85ffcf7056cea237e13